// quick-erd/dist/core/meta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tableNameToLine = exports.tableNameToRegex = exports.tableNameRegex_g = exports.tableNameRegex = exports.viewToLine = exports.viewPositionRegex = exports.viewLineRegex = exports.zoomToLine = exports.zoomValueRegex = exports.zoomLineRegex = void 0;
exports.zoomLineRegex = /# zoom: [0-9.]+/;
exports.zoomValueRegex = /# zoom: ([0-9.]+)/;
function zoomToLine(zoom) {
    return `# zoom: ${zoom.toFixed(3)}`;
}
exports.zoomToLine = zoomToLine;
exports.viewLineRegex = /# view: \([0-9-]+, [0-9-]+\)/;
exports.viewPositionRegex = /# view: \(([0-9-.]+), ([0-9-.]+)\)/;
function viewToLine(view) {
    const x = view.x.toFixed(0);
    const y = view.y.toFixed(0);
    return `# view: (${x}, ${y})`;
}
exports.viewToLine = viewToLine;
exports.tableNameRegex = /# (\w+) \(([0-9-.]+), ([0-9-.]+)\)/;
exports.tableNameRegex_g = new RegExp(exports.tableNameRegex.toString().slice(1, -1), 'g');
function tableNameToRegex(name) {
    return new RegExp(`# ${name} \\([0-9-]+, [0-9-]+\\)`);
}
exports.tableNameToRegex = tableNameToRegex;
function tableNameToLine(name, position) {
    const x = position.x.toFixed(0);
    const y = position.y.toFixed(0);
    return `# ${name} (${x}, ${y})`;
}
exports.tableNameToLine = tableNameToLine;
